File: FSXScript/FsxScriptResourceFormatSaver.fs
namespace FSXScript

open Godot
open Godot.Bridge
open Godot.Collections

type FsxScriptResourceFormatSaver() as self =
    inherit ResourceFormatSaver()
    
    override _._GetRecognizedExtensions(resource) =
        let array = new PackedStringArray()
        
        if resource.IsClass(nameof(FsxScript)) then
            array.Add("fsx")
        
        array
    
    override _._Recognize(resource) =
        printf($"{resource}")
        resource.IsClass(nameof(FsxScript))
            
    override _._RecognizePath(resource, path) =
        printf($"{resource} {path}")
        path.EndsWith(".fsx") && self._Recognize(resource)
    
    override _._Save(resource, path, flags) =
        let file = FileAccess.Open(path, FileAccess.ModeFlags.Write)
        resource.SetPath(path)
        let script = resource :?> Script
        file.StoreString(script.GetSourceCode())
        file.Close()        
        Error.Ok
        
    override _._SetUid(path, uid) =
        Error.Unconfigured
        
    static member BindMethods(context: ClassDBRegistrationContext) =
        context.BindConstructor(fun () -> new ResourceFormatSaver())

File: fsx-script/src/fsx_script_resource_format_saver.rs
use godot::classes::{FileAccess, IResourceFormatSaver, ResourceFormatSaver};
use godot::classes::file_access::ModeFlags;
use godot::global::Error;
use godot::meta::AsArg;
use godot::prelude::*;

use crate::fsx_script::FsxScript;

#[derive(GodotClass)]
#[class(base = ResourceFormatSaver)]
pub(crate) struct FsxScriptResourceFormatSaver {
    base: Base<ResourceFormatSaver>,
}

#[godot_api]
impl IResourceFormatSaver for FsxScriptResourceFormatSaver {
    fn init(base: Base<Self::Base>) -> Self {
        Self { base }
    }

    fn save(&mut self, resource: Option<Gd<Resource>>, path: GString, _flags: u32) -> Error {
        let resource = match resource {
            None => {return Error::ERR_INVALID_PARAMETER}
            Some(resource) => { resource }
        };
        let script = resource.cast::<FsxScript>();
        let file = FileAccess::open(path.into_arg(), ModeFlags::WRITE);
        match file {
            None => {
                godot_error!("Could not open {path} for writing");
                FileAccess::get_open_error()
            }
            Some(mut file) => {
                file.store_string(script.get_source_code().into_arg());
                let error = file.get_error();
                file.close();
                error
            }
        }
    }

    fn recognize(&self, resource: Option<Gd<Resource>>) -> bool {
        match resource {
            None => false,
            Some(resource) => resource.is_class(GString::from("Script").into_arg()),
        }

    }

    fn get_recognized_extensions(&self, resource: Option<Gd<Resource>>) -> PackedStringArray {
        if self.recognize(resource) {
            PackedStringArray::from(&[GString::from("fsx")])
        } else {
            PackedStringArray::new()
        }
    }
}
